### PR TITLE
enhancement_template: add API extension sections

### PR DESCRIPTION
**Pull request by sttts**
Implement the enhancement template part of https://github.com/openshift/enhancements/pull/908.

### CODE REVIEW

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md]: added

File: guidelines/enhancement_template.md
Comment (bparees) [inline guidelines/enhancement_template.md:129]: would you also expect new resource types, and/or new fields being added to existing resource types, to be described here?

(I would, though they also tend to end up in the implementation details section currently)

File: guidelines/enhancement_template.md
Comment (bparees) [inline guidelines/enhancement_template.md:123]: part of previous sentence?

File: guidelines/enhancement_template.md
Comment (bparees) [inline guidelines/enhancement_template.md:283]: this seems very focused on the case where a new api resource type is being added, but doesn't cover the case where someone is trying to extend an existing api with new fields or behavior.

how are "api extensions" defined?

1) new resource type?
2) new field on a resource type?
3) new webhook on an existing type (or being added as part of adding a new resource type)?

File: guidelines/enhancement_template.md
Comment (bparees) [inline guidelines/enhancement_template.md:286]: this could use a more explicit example for people to follow, e.g. "when api extension $foo was added, the following SLI was introduced...."

if i'm adding a new config resource for builds(for example), i am not clear what it would mean to add a new SLI for that resource, or if such a thing is even really applicable to that api change.

File: guidelines/enhancement_template.md
Comment (bparees) [reply guidelines/enhancement_template.md:283]: actually i guess it's focused on the case where a new webhook or finalizer is being added?

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [inline guidelines/enhancement_template.md:307]: Throughout can we follow the existing text formatting of beginning lines/sentences with a capital letter? Makes template consistent and more readable.

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [inline guidelines/enhancement_template.md]:
```suggestion
- Name the API extensions (CRDs, webhooks, aggregated API servers) this enhancement adds or modifies
```
Capitalize throughout

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md:12]: Would like to link to the list of possible approvers. https://github.com/openshift/api/blob/master/OWNERS is not that. An OWNERS_ALIASES file might work as in upstream.

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [inline guidelines/enhancement_template.md:120]: Q: why is this section so far from the Impact of API Extensions section? Is there a reason to seperate them? Does doing so impact readability?

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [inline guidelines/enhancement_template.md:311]:
```suggestion
- Describe which OCP teams are likely to be called upon in case of escalation with one of the failure modes
```

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [inline guidelines/enhancement_template.md:312]:
```suggestion
  and add them as reviewers to this enhancement.
```

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [reply guidelines/enhancement_template.md:311]: I really like this btw

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [inline guidelines/enhancement_template.md:125]:
```suggestion
  by other parties than the authoring team (including upstream resources), and, if yes, how?
```

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [inline guidelines/enhancement_template.md:127]: Do you want to add a specific example of something owned by someone else and how you'd like to see that described?

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md:120]: the impact is in the details part, the listing is here, for easier reviewing.

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md:127]: the examples below are like that. A finalizer on namespaces modifies a kube object. A label restriction modifies another object normal behaviour.

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md:302]: the impact on the SLI, shown above. Will clarify.

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md:307]: done

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md:308]: there might be behaviours, like slowing down requests or adding finalizers, that influence cluster health.

File: guidelines/enhancement_template.md
Comment (soltysh) [reply guidelines/enhancement_template.md:286]: As stated above `API Extensions` is everything impacting APIs: webhooks, finalizers, CRDs. 
The effect of a new API field being added should be describe throughout the enhancement discussing that kind of extension of the existing API, no?

**Review by soltysh — APPROVED** [review]
this lgtm

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md]: done

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md]: done

File: guidelines/enhancement_template.md
Comment (bparees) [reply guidelines/enhancement_template.md]: hm, I don't see why "who owns the api" vs "who is writing the EP" matters?  That may affect who needs to review it, but not what the design needs to cover.

what matters is the kind of change you are proposing to the product, and the effects it may have that need to be considered.

For which it seems like "api" changes need to be broken down into a few categories, e.g.:

1) new api type
2) updating existing api type (e.g. new fields)
3) adding a webhook (not sure if conversion vs admission need to be divided)
4) adding/changing validations on an apitype (adding is usually not allowed, of course)
5) adding/changing rbac associated w/ an api type

i'm not saying you need to cover all of them in your changes here, but i think it's important we make it clear which kinds of changes these new sections are applicable to.

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [reply guidelines/enhancement_template.md:129]: Q: if it's owned by someone else do you want them to explicitly say it like you requested like in failure modes below? Should they be added as reviewers?

Comment (sttts) [reply guidelines/enhancement_template.md:129]: I would leave that to the API review to judge. Some APIs have owners that can be added. If the modification is more generic, i.e. to multiple resources, it's not that clear.

File: guidelines/enhancement_template.md
Comment (dhellmann) [reply guidelines/enhancement_template.md:12]: We're planning to change the way we list reviewers generally, to include the scope of review being requested and other details to help the final approver ensure that all areas are covered. I think we can address the question of how to find people for different types of reviews as part of that change.

File: guidelines/enhancement_template.md
Comment (kikisdeliveryservice) [reply guidelines/enhancement_template.md:130]: lgtm I just don't have perms to resolve my comment :)

File: guidelines/enhancement_template.md
Comment (sttts) [reply guidelines/enhancement_template.md:123]: CRDs certainly have less operation impact. But listing them is a good idea IMO.

I fear a little that we have to try to keep the balance. We can add more and more to stop certain common mistakes from happening. But the risk is that ppl neglect certain sections because there are just too many. I notice this a lot with the risk, test and version skew section at the bottom. Often they are not seriously filled in.

Comment (bparees) [reply guidelines/enhancement_template.md:123]: can we give more guidance about when those sections are applicable?  my fear is that someone adding a new CRD is going to have no idea what to put in the "SLI" section (i don't really know what we'd expect them to put there)

**Pull request merged**
